### PR TITLE
New type next_url

### PR DESCRIPTION
**Pull request by priyankjalu89**
image

### CODE REVIEW

File: tcl/cast.tcl [hunk 1]
```diff
@@ -853,4 +853,27 @@ namespace eval qc::cast {
...
+        }
+
+        # check for malicious mal-formed url
+        if { ![qc::is url $next_url] } {
+            return -code error -errorcode CAST "Could not cast \"[html_escape $next_url]\" to valid next_url. Invalid URL."
+        }
+        
+        # check that redirection is to the same domain
+        set conn_host [qc::conn_host]
+        if { ![regexp "^https?://${conn_host}(:\[0-9\]+)?(\\?|/|\$)" $next_url] } {
```
Comment (daniel-qcode) [inline tcl/cast.tcl:873]: do we also need to handle https://www.abc.com#test

File: tcl/cast.tcl [hunk 2]
```diff
@@ -853,4 +853,27 @@ namespace eval qc::cast {
...
+
+        # check for malicious mal-formed url
+        if { ![qc::is url $next_url] } {
+            return -code error -errorcode CAST "Could not cast \"[html_escape $next_url]\" to valid next_url. Invalid URL."
+        }
+        
+        # check that redirection is to the same domain
+        set conn_host [qc::conn_host]
+        if { ![regexp "^https?://${conn_host}(:\[0-9\]+)?(\\?|#|/|\$)" $next_url] } {
+            return -code error -errorcode CAST "Could not cast \"[html_escape $next_url]\" to valid next_url. Domain must be \"$conn_host\"."
```
Comment (davidqc) [inline tcl/cast.tcl:874]: $conn_host can be controlled by header manipulation so is not completely safe - does this end up being rendered back to the user? Is so, it need escaping (Maybe something further up takes care of that?)

Comment (priyankjalu89) [reply tcl/cast.tcl:874]: Yes this end up being rendered back to the user and yes I think it should be the caller's responsibility to take care of that. Would it be okay to escape the next_url before it is passed to this cast api [here](https://github.com/qcode-software/qcode-tcl/pull/715/files#diff-bb1f05244dfece6f46f7e54ddff8d5a419586b13eca9f5294fc8a3229884a27fR135)

**Comment by davidqc** [comment]
Can  we use ```qc::cast next_url``` within ```qc::response action redirect```?

File: tcl/cast.tcl (hunk 2 above)
Comment (davidqc) [reply tcl/cast.tcl:874]: We're pulling ```conn_host``` directly out of the headers ```[qc::conn_host]```within this proc so the caller can't deal with it.

Comment (priyankjalu89) [reply tcl/cast.tcl:874]: Good point. I can `html_escape` the `string` at the start of the API if that helps ?

Comment (priyankjalu89) [reply tcl/cast.tcl:874]: If we cannot trust the `conn_host` the only other option I can think of is using a `param`.

**Comment by priyankjalu89** [comment]
> Can we use `qc::cast next_url` within `qc::response action redirect`? Unsure whether the errors presented to the user would become unsuitable.

Casting is something we do internally so providing and end user with a message e.g `could not cast <x> to Boolean` would not be a helpful message and it also confuse them as to what they should do next. `Invalid value` on the other hand lets them know that they have entered something which is not a valid input and encourage them correct it.

**Review by davidqc — APPROVED** [review]
OK if we can't reuse the ```cast next_url``` code in ```response action redirect/login```

**Comment by davidqc** [comment]
image

**Pull request merged**